src/frontends/lean/builtin_cmds.cpp

```diff
@@ -403,13 +411,22 @@ static environment eval_cmd(parser & p) {
     auto out = p.mk_message(p.cmd_pos(), p.pos(), INFORMATION);
     out.set_caption("eval result");
     scope_traces_as_messages scope_traces(p.get_stream_name(), p.cmd_pos());
-    char * redirected; size_t redir_size;
-    FILE * fp = open_memstream(&redirected, &redir_size);
-    std::unique_ptr<char *, stream_buffer_delete> stream_buffer(&redirected);
+    object_ref pipe(lean_io_mk_pipe(box(0)));
+    if (io_result_is_error(pipe.raw())) {
+        message_builder msg = p.mk_message(p.cmd_pos(), p.pos(), ERROR);
+        object * err = io_result_get_error(pipe.raw());
+        inc_ref(err);
+        object * str = lean_io_error_to_string(err);
+        msg << string_to_std(str);
+        msg.report();
+        dec_ref(str);
+        return p.env();
+    }
+    pipe = object_ref(lean_ctor_get(pipe.raw(), 0), true);
     object_ref r;
-
     try {
-        object_ref new_fp(lean_io_wrap_handle(fp));
+        object_ref new_fp(lean_ctor_get(pipe.raw(), 1), true);
+        pipe = object_ref(lean_ctor_get(pipe.raw(), 0), true);
```

Comment:
I am confused by this line. You are accessing the first field `pipe.input`, which has type `FS.Handle`. Then, you overwrite the `pipe` reference with it. Now, the `pipe` is not a `pipe` anymore, but a `FS.Handle`.  Is this intentional or a typo?

Reply:
This is intentional. I could make it clearer and create a new variable to hold that `FS.Handle`. I wanted to make sure to release the `pipe` itself so that the writing handle will get closed at the end of the current scope. But you're right, it  could be clearer.